### ISSUE TITLE
[feature request] Including NADAM optimizer in torch.optim

### ISSUE THREAD

**Issue by anubane** [pytorch/pytorch#63391]
## Issue description

`torch.optim` has most of the useful optimizers, but for some reason, NADAM is missing. This ticket is a request to include the NADAM optimizer. I have included an implementation that is based on the ADAM optimizer already present. Kindly verify and add if feasible. Created this ticket because I do not see the NADAM optimizer on the page: https://pytorch.org/docs/stable/optim.html

## Code example

```Python
class NadamOpt(Optimizer):
    """
    Borrows from http://cs229.stanford.edu/proj2015/054_report.pdf
    and https://medium.com/konvergen/modifying-adam-to-use-nesterov-accelerated-gradients-nesterov-accelerated-adaptive-moment-67154177e1fd
    """
    def __init__(self, params,  lr=1e-3, betas=(0.9, 0.999), eps=1e-8, weight_decay=0, amsgrad=False):
        defaults = dict(lr=lr, betas=betas, eps=eps, weight_decay=weight_decay)
        super(NadamOpt, self).__init__(params, defaults)

    @torch.no_grad()
    def step(self):
        for group in self.param_groups:
            params_with_grad = []
            grads = []
            exp_avgs = []       # for gradients
            exp_avgs_cap = []       # for gradients - lookahead
            exp_avg_sqs = []    # for squared gradients
            max_exp_avg_sqs = []
            state_steps = []
            beta1, beta2 = group['betas']

            for p in group['params']:
                if p.grad is not None:
                    params_with_grad.append(p)
                    grads.append(p.grad)

                    state = self.state[p]
                    # Lazy state initialization
                    if len(state) == 0:
                        state['step'] = 0
                        # Exponential moving average of gradient values
                        state['exp_avg'] = torch.zeros_like(p, memory_format=torch.preserve_format)
                        # Exponential moving average of squared gradient values
                        state['exp_avg_sq'] = torch.zeros_like(p, memory_format=torch.preserve_format)

                    exp_avgs.append(state['exp_avg'])
                    exp_avg_sqs.append(state['exp_avg_sq'])

                    # update the steps for each param group update
                    state['step'] += 1
                    # record the step after step update
                    state_steps.append(state['step'])
            
            # NADAM Optimizer
            for i, param in enumerate(params_with_grad):
                grad = grads[i]
                exp_avg = exp_avgs[i]
                exp_avg_sq = exp_avg_sqs[i]
                step = state_steps[i]

                bias_correction1 = 1 - beta1 ** step
                bias_correction2 = 1 - beta2 ** step

                if group['weight_decay'] != 0:
                    grad = grad.add(param, alpha=group['weight_decay'])

                # Decay the first and second moment running average coefficient
                exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
                exp_avg_sq.mul_(beta2).addcmul_(grad, grad.conj(), value=1 - beta2)
                
                exp_avg.mul_(1 / bias_correction1)
                exp_avg_sq.mul_(1 / bias_correction2)

                exp_avg_cap = torch.mul(exp_avg, beta1).add_(grad, alpha=1 - beta1)

                denom = exp_avg_sq.sqrt().add_(group['eps'])

                param.addcdiv_(exp_avg_cap, denom, value=-group['lr'])

```

The part marked as `#NADAM Optimizer` can be moved to the `_functional.py` as `nadam()` and a call to `F.nadam()` can replace it here.

Looking forward to the response and critique of this idea!

cc @vincentqb @iramazanli

**Comment by anubane** [comment]
The proposed implementation is better than the one suggested in issue [https://github.com/pytorch/pytorch/pull/59009](#59009)

**Comment by iramazanli** [comment]
Thanks for this amazing suggestion. This issue actually got solved in #59009

**Comment by anubane** [comment]
> Thanks for this amazing suggestion. This issue actually got solved in #59009

@iramazanli I do not see how the issue is resolved. PyTorch still does not have an official NADAM implementation!

**Comment by anubane** [comment]
As of writing this, I do not see the NAdam optimizer on this page:
https://pytorch.org/docs/stable/optim.html#algorithms
I contest the closing of this issue. ML developers either have to write
their own implementation or use third party implementation.
It would be beneficial to have an official implementation!
